Windows: display which program is causing port conflict

cwillisf opened the pull request:
### Proposed Changes

On Windows only (for now): if Scratch Link fails to bind its communication port, try to display the process ID and name for any process using that port.

Before:
![image](https://user-images.githubusercontent.com/7019101/84822319-55290a00-afd1-11ea-9666-1890cffc377a.png)

After:
![image](https://user-images.githubusercontent.com/7019101/84822431-7be74080-afd1-11ea-9fbf-1fa8d2a4b524.png)

Note that `FiddlerWinsock.cs` was written by another group and should be treated as an external dependency. Also, `AddressInUse.Designer.cs` and `AddressInUse.resx` are auto-generated code created by the dialog editor tool in Visual Studio.

### Reason for Changes

Under normal circumstances a particular port can only be used by one application at a time. If Scratch Link fails to bind its port, the most likely reason is that another application has already bound that port. Previously, Scratch Link would report this but figuring out which application is causing the conflict requires running commands at an administrator-level command prompt, which is at best difficult to explain and at worst impossible for users with restricted permissions.

This feature should help users figure out which application(s) are conflicting with Scratch Link without needing to use a command prompt.

Based on help desk tickets, this seems to affect Windows users much more often than macOS users. I still intend to implement this on macOS in a separate change.

Commit:
try to display process info when port is in use

Commit:
add button to copy address-in-use dialog details

fsih commented on Windows/scratch-link/Dialogs/AddressInUse.resx:
```diff
@@ -0,0 +1,120 @@
+    <data name="Bitmap1" mimetype="application/x-microsoft.net.object.binary.base64">
+        <value>[base64 mime encoded serialized .NET Framework object]</value>
+    </data>
+    <data name="Icon1" type="System.Drawing.Icon, System.Drawing" mimetype="application/x-microsoft.net.object.bytearray.base64">
+        <value>[base64 mime encoded string representing a byte array form of the .NET Framework object]</value>
+        <comment>This is a comment</comment>
+    </data>
+                
+    There are any number of "resheader" rows that contain simple 
+    name/value pairs.
+    
+    Each data row contains a name, and value. The row also contains a 
+    type or mimetype. Type corresponds to a .NET class that support 
+    text/value conversion through the TypeConverter architecture. 
+    Classes that don't support this are serialized and stored with the 
+    mimetype set.
+    
+    The mimetype is used for serialized objects, and tells the 
+    ResXResourceReader how to depersist the object. This is currently not 
+    extensible. For a given mimetype the value must be set accordingly:
+    
+    Note - application/x-microsoft.net.object.binary.base64 is the format 
+    that the ResXResourceWriter will generate, however the reader can 
+    read any of the formats listed below.
+    
+    mimetype: application/x-microsoft.net.object.binary.base64
+    value   : The object must be serialized with 
+            : System.Runtime.Serialization.Formatters.Binary.BinaryFormatter
+            : and then encoded with base64 encoding.
+    
+    mimetype: application/x-microsoft.net.object.soap.base64
+    value   : The object must be serialized with 
+            : System.Runtime.Serialization.Formatters.Soap.SoapFormatter
+            : and then encoded with base64 encoding.
+
+    mimetype: application/x-microsoft.net.object.bytearray.base64
+    value   : The object must be serialized into a byte array 
+            : using a System.ComponentModel.TypeConverter
+            : and then encoded with base64 encoding.
+    -->
```
Seems unnecessary

fsih replied:
Overlapping

fsih replied:
comments

fsih replied:
out of order

fsih replied:
I don't like the letter e

fsih replied:
please remove all e's

BryceLTaylor commented on Windows/scratch-link/Dialogs/AddressInUse.cs:
```diff
@@ -0,0 +1,51 @@
+                var item = new ListViewItem(new [] { process.ProcessName, pid.ToString() });
+                listView1.Items.Add(item);
+                textContent += pid.ToString() + '\t' + process.ProcessName + '\n';
+            }
```
We should add the label2 text here as an else so we only show it if we haven't found another process.

this.label2.Text = "Some details may be missing if this application is not running as Administrator.";

BryceLTaylor commented on Windows/scratch-link/Dialogs/AddressInUse.cs:
```diff
@@ -0,0 +1,51 @@
+            if (pid > 0)
+            {
+                var process = Process.GetProcessById(pid);
+                var item = new ListViewItem(new [] { process.ProcessName, pid.ToString() });
```
Should this be a list?  Doesn't look like a list.

BryceLTaylor reviewed (dismissed):
Mostly looks good.  See comments.

fsih commented on Windows/scratch-link/Dialogs/AddressInUse.resx:
```diff
@@ -0,0 +1,120 @@
+    value   : The object must be serialized with 
+            : System.Runtime.Serialization.Formatters.Binary.BinaryFormatter
+            : and then encoded with base64 encoding.
+    
+    mimetype: application/x-microsoft.net.object.soap.base64
+    value   : The object must be serialized with 
+            : System.Runtime.Serialization.Formatters.Soap.SoapFormatter
+            : and then encoded with base64 encoding.
+
+    mimetype: application/x-microsoft.net.object.bytearray.base64
```
```suggestion
    
hello
```

fsih replied:
```suggestion
            other hello
```

fsih commented on Windows/scratch-link/Dialogs/AddressInUse.resx:
```diff
@@ -0,0 +1,120 @@
+    value   : The object must be serialized with 
+            : System.Runtime.Serialization.Formatters.Binary.BinaryFormatter
+            : and then encoded with base64 encoding.
+    
+    mimetype: application/x-microsoft.net.object.soap.base64
+    value   : The object must be serialized with 
+            : System.Runtime.Serialization.Formatters.Soap.SoapFormatter
+            : and then encoded with base64 encoding.
+
+    mimetype: application/x-microsoft.net.object.bytearray.base64
+    value   : The object must be serialized into a byte array 
+            : using a System.ComponentModel.TypeConverter
+            : and then encoded with base64 encoding.
```
```suggestion
    
    mimetype: application/x-microsoft.net.object.soap.base64
    value   : The object must be serialized with 
            : System.Runtime.Serialization.Formatters.Soap.SoapFormatter
            : and then encoded with base64 encoding.

    mimetype: application/x-microsoft.net.object.bytearray.base64
    value   : The object must be serialized into a byte array 
            : using a System.ComponentModel.TypeConverter
            : and then encoded with base64 encoding.
```

Commit:
more process details, hide warning unless relevant

cwillisf commented:
I renamed "label2" to "labelNoResults" to be a little more descriptive, and now I set its visibility based on whether or not `MapLocalPortToProcessId` returned a good PID. If so, the new code displays more details than it used to including the title of the main window (if any) and the file name.

Here's what it looks like now (note that Scratch Link has no main window):
![image](https://user-images.githubusercontent.com/7019101/84959500-4a937100-b0b4-11ea-8b4c-76b8c96923a6.png)

...and here's the corresponding "Copy" text:
```
Scratch Link was unable to start because port 20110 is already in use.

This means Scratch Link is already running or another application is using that port.

This application will now exit.

Details:
Process Name:	ScratchLink
Window Title:	
File Name:	C:\Program Files (x86)\Scratch Link\ScratchLink.exe
Process ID:	34692
TCP Port:	20110
```

cwillisf commented:
Oh, and here's what it looks like if there's a failure in trying to determine which program is using the port:
![image](https://user-images.githubusercontent.com/7019101/84959727-c5f52280-b0b4-11ea-9974-55a21cf64882.png)

Corresponding "Copy" text:
```
Scratch Link was unable to start because port 20110 is already in use.

This means Scratch Link is already running or another application is using that port.

This application will now exit.

Details:
Could not collect details. Running as Administrator might help.
```

fsih reviewed (approved):
LGTM!

![image](https://user-images.githubusercontent.com/7019101/84959500-4a937100-b0b4-11ea-8b4c-76b8c96923a6.png)
One last thought is that in this particular case, it's not clear that the details area is referring to the process that's preventing Scratch Link from opening. It looks like it might just be telling you about Scratch Link itself, the process that just quit?

cwillisf commented:
Ooh, good point about the confusion. Hmmm...

Commit:
make details label more descriptive

When the conflicting program is another instance of Scratch Link, just
using "Details" as the label could be confusing: is it showing details
about the current program or the conflicting program? This change will
hopefully make it clear that the "Details about the conflicting program"
are about the other instance.

fsih reviewed (approved):
That seems like a good change!
